File: iosApp/iosApp/Modules/CurrencyReceive/CurrencyReceiveWireframeFactory.swift
// Created by web3d4v on 13/06/2022.
// Copyright (c) 2022 Sons Of Crypto.
// SPDX-License-Identifier: MIT

import UIKit
import web3walletcore

// MARK: - CurrencyReceiveWireframeFactory

protocol CurrencyReceiveWireframeFactory {
    func make(
        _ parent: UIViewController?,
        context: CurrencyReceiveWireframeContext
    ) -> CurrencyReceiveWireframe
}

// MARK: - DefaultCurrencyReceiveWireframeFactory

final class DefaultCurrencyReceiveWireframeFactory {
    private let networksService: NetworksService

    init(networksService: NetworksService) {
        self.networksService = networksService
    }
}

extension DefaultCurrencyReceiveWireframeFactory: CurrencyReceiveWireframeFactory {

    func make(
        _ parent: UIViewController?,
        context: CurrencyReceiveWireframeContext
    ) -> CurrencyReceiveWireframe {
        DefaultCurrencyReceiveWireframe(
            parent,
            context: context,
            networksService: networksService
        )
    }
}

// MARK: - Assembler

final class CurrencyReceiveWireframeFactoryAssembler: AssemblerComponent {
    
    func register(to registry: AssemblerRegistry) {
        registry.register(scope: .instance) { resolver -> CurrencyReceiveWireframeFactory in
            DefaultCurrencyReceiveWireframeFactory(
                networksService: resolver.resolve()
            )
        }
    }
}

File: androidApp/src/main/java/com/sonsofcrypto/web3wallet/android/modules/compose/currencyreceive/CurrencyReceiveWireframeFactory.kt
package com.sonsofcrypto.web3wallet.android.modules.compose.currencyreceive

import androidx.fragment.app.Fragment
import com.sonsofcrypto.web3lib.services.networks.NetworksService
import com.sonsofcrypto.web3lib.utils.WeakRef
import com.sonsofcrypto.web3wallet.android.common.AssemblerComponent
import com.sonsofcrypto.web3wallet.android.common.AssemblerRegistry
import com.sonsofcrypto.web3wallet.android.common.AssemblerRegistryScope
import com.sonsofcrypto.web3walletcore.modules.currencyReceive.CurrencyReceiveWireframe
import com.sonsofcrypto.web3walletcore.modules.currencyReceive.CurrencyReceiveWireframeContext
import smartadapter.internal.extension.name

interface CurrencyReceiveWireframeFactory {
    fun make(parent: Fragment?, context: CurrencyReceiveWireframeContext): CurrencyReceiveWireframe
}

class DefaultCurrencyReceiveWireframeFactory(
    private val networksService: NetworksService,
): CurrencyReceiveWireframeFactory {

    override fun make(
        parent: Fragment?, context: CurrencyReceiveWireframeContext
    ): CurrencyReceiveWireframe = DefaultCurrencyReceiveWireframe(
            parent?.let { WeakRef(it) },
            context,
            networksService
        )
    }

class CurrencyReceiveWireframeFactoryAssembler: AssemblerComponent {

    override fun register(to: AssemblerRegistry) {
        to.register(CurrencyReceiveWireframeFactory::class.name, AssemblerRegistryScope.INSTANCE) {
            DefaultCurrencyReceiveWireframeFactory(
                it.resolve(NetworksService::class.name)
            )
        }
    }
}